src/plugins/hetero/src/compiled_model.cpp

```diff
@@ -0,0 +1,762 @@
+                    if (affinities[node.get()] != affinities[InputNode(input)]) {
+                        subgraphInputs.insert(input);
+                    }
+                }
+            }
+        }
+
+        // Assign each node subgraph ID
+        auto CollectSubgraphs = [&] {
+            std::deque<int> subgraphIds;
+            NodeMap<int*> subgraphIdPtrs;
+            for (auto&& node : orderedOps) {
+                auto allNodeInputs = node->inputs();
+                std::vector<Input> inputs;
+                for (auto&& input : allNodeInputs) {
+                    if (!InferenceEngine::details::contains(subgraphInputs,
+                                                            input)) {  // TODO vurusovs REPLACE with ov::util::contains
+                        inputs.emplace_back(std::move(input));
+                    }
+                }
+                if (inputs.empty()) {
+                    subgraphIds.push_back(static_cast<int>(subgraphIds.size()));
+                    subgraphIdPtrs.emplace(node.get(), &(subgraphIds.back()));
+                } else {
+                    auto firstInputSubgraphIdPtr = subgraphIdPtrs[InputNode(inputs.front())];
+                    for (auto&& input : inputs) {
+                        auto inputId = *subgraphIdPtrs[InputNode(input)];
+                        for (auto& subgraphId : subgraphIds) {
+                            if (subgraphId == inputId) {
+                                subgraphId = *firstInputSubgraphIdPtr;
+                            }
+                        }
+                    }
+                    subgraphIdPtrs.emplace(node.get(), firstInputSubgraphIdPtr);
+                }
+            }
+            NodeMap<int> result;
+            for (auto&& subgraphIdPtr : subgraphIdPtrs) {
+                result.emplace(subgraphIdPtr.first, *(subgraphIdPtr.second));
+            }
+            return result;
+        };
+
+        // Split cyclic dependencies.
+        for (std::size_t prevSubgraphs = 0, cyclicSplitStep = 0; prevSubgraphs != subgraphInputs.size();
+             ++cyclicSplitStep) {
+            OPENVINO_ASSERT(cyclicSplitStep < orderedOps.size());
+            prevSubgraphs = subgraphInputs.size();
+            auto subgraphIds = CollectSubgraphs();
+            // All inputs that belong to the same subgraph as node
+            std::unordered_map<ov::Node*, InputSet> nodeSubgraphInputDependencies;
+            // All inputs that depends on the same subgraph as node
+            std::unordered_map<ov::Node*, InputSet> nodeSubgraphCyclicInputDependencies;
+            for (auto&& node : orderedOps) {
+                auto& nodeSubgraphInputDependency = nodeSubgraphInputDependencies[node.get()];
+                auto allNodeSubgraphInputs =
+                    InferenceEngine::details::Intersection(nodeInputDependencies[node.get()], subgraphInputs);
```

Comment:
I think we need avoid legacy API in new plugins...

Reply:
https://github.com/openvinotoolkit/openvino/pull/18392#discussion_r1259437222